### PR TITLE
Inital support for Azure VMSS

### PR DESCRIPTION
**Pull request by johscheuer**
This PR adds support for VMSS in Azure. This PR is based on this closed PR: https://github.com/prometheus/prometheus/pull/2376

Solves the following issue: https://github.com/prometheus/prometheus/issues/2369

### CODE REVIEW

**Comment by brian-brazil** [comment]
Thanks for your PR, I'm afraid we're not accepting any new SD mechanisms currently as you can see from https://github.com/prometheus/prometheus/blob/master/discovery/README.md#service-discovery. I recommend integrating via file_sd for now.

**Comment by johscheuer** [comment]
Hey, actually this isn't a new SD mechanism instead it's an extension to the existing Azure SD configuration.

**Comment by brian-brazil** [comment]
Hmm, I'll reopen then. I can't guarantee anyone will get to reviewing this, and #3987 indicates there may be bigger issues with this SD.

**Comment by johscheuer** [comment]
We should definitly update the Azure dependencies to the latetst stable version. I can so this in this PR (or create a second one)

**Comment by johscheuer** [comment]
I would wait until https://github.com/prometheus/prometheus/pull/4147 is merged and then rebase my PR

**Comment by sylr** [comment]
@johscheuer I made https://github.com/sylr/prometheus/tree/azure-vmss which is this PR rebased on top of https://github.com/sylr/prometheus/tree/azure-autorest which is #4147 rebased on top of https://github.com/prometheus/prometheus/tree/release-2.3

Building this branch fails :

```
Step 5/20 : RUN make build
 ---> Running in 01e0f9b949b4
GOOS= GOARCH= go get -u github.com/prometheus/promu
>> building binaries
/go/bin/promu build --prefix /go/src/github.com/prometheus/prometheus
 >   prometheus
# github.com/prometheus/prometheus/discovery/azure
discovery/azure/azure.go:179:20: cannot use spt (type *adal.ServicePrincipalToken) as type autorest.Authorizer in assignment:
        *adal.ServicePrincipalToken does not implement autorest.Authorizer (missing WithAuthorization method)
discovery/azure/azure.go:182:22: cannot use spt (type *adal.ServicePrincipalToken) as type autorest.Authorizer in assignment:
        *adal.ServicePrincipalToken does not implement autorest.Authorizer (missing WithAuthorization method)
!! command failed: build -o /go/src/github.com/prometheus/prometheus/prometheus -ldflags -X github.com/prometheus/prometheus/vendor/github.com/prometheus/common/version.Version=2.3.1 -X github.com/prometheus/prometheus/vendor/github.com/prometheus/common/version.Revision=27a01f65b6ebffd8d208687c37141c9c6d84f2b4 -X github.com/prometheus/prometheus/vendor/github.com/prometheus/common/version.Branch=HEAD -X github.com/prometheus/prometheus/vendor/github.com/prometheus/common/version.BuildUser=root@01e0f9b949b4 -X github.com/prometheus/prometheus/vendor/github.com/prometheus/common/version.BuildDate=20180704-11:45:25  -extldflags '-static' -a -tags netgo github.com/prometheus/prometheus/cmd/prometheus: exit status 2
make: *** [Makefile.common:82: build] Error 1
The command '/bin/sh -c make build' returned a non-zero code: 2
```

**Comment by johscheuer** [comment]
I can take a look later on you probably need to adjust the authentication method since the underlying Azure API has changed. As I said as soon as https://github.com/prometheus/prometheus/pull/4147 is merged I will rebase my PR and add the according fixes (should only be minor changes).

**Comment by marratj** [comment]
@johscheuer I was searching for a way to integrate Managed Service Identity authentication for the Azure SD and found that the vendored API is way older than the official Azure one.

Good to know that you will be working on updating the Azure dependencies 👍

**Comment by johscheuer** [comment]
Rebased since https://github.com/prometheus/prometheus/pull/4147 is merged

**Comment by johscheuer** [comment]
In a following PR I would update the Azure go SDK (Prometheus currently uses a relative old version of the go SDK).

File: discovery/azure/azure.go [hunk 1]
```diff
@@ -180,16 +190,31 @@ type azureResource struct {
 	ResourceGroup string
 }
 
+// virtualMachine represents an Azure virtual machine (which can also be created by a VMSS)
+type virtualMachine struct {
+	ID             *string
```
Comment (brian-brazil) [inline discovery/azure/azure.go]: Why are these all pointers?

File: discovery/azure/azure.go [hunk 2]
```diff
@@ -212,22 +237,26 @@ func (d *Discovery) refresh() (tg *targetgroup.Group, err error) {
...
+		return tg, fmt.Errorf("could not get virtual machines: %s", err)
 	}
-	machines = append(machines, *result.Value...)
 
-	// If we still have results, keep going until we have no more.
-	for result.NextLink != nil {
-		result, err = client.vm.ListAllNextResults(result)
+	level.Debug(d.logger).Log("Found %d virtual machines during Azure discovery.", len(machines))
+
+	// Load the vms managed by scale sets
```
Comment (brian-brazil) [inline discovery/azure/azure.go]: Full stop

File: discovery/azure/azure.go [hunk 3]
```diff
@@ -212,22 +237,26 @@ func (d *Discovery) refresh() (tg *targetgroup.Group, err error) {
...
 	if err != nil {
-		return tg, fmt.Errorf("could not list virtual machines: %s", err)
+		return tg, fmt.Errorf("could not get virtual machines: %s", err)
 	}
-	machines = append(machines, *result.Value...)
 
-	// If we still have results, keep going until we have no more.
-	for result.NextLink != nil {
-		result, err = client.vm.ListAllNextResults(result)
+	level.Debug(d.logger).Log("Found %d virtual machines during Azure discovery.", len(machines))
```
Comment (brian-brazil) [inline discovery/azure/azure.go]: This isn't following the usual go-kit way of doing structured logging.

File: discovery/azure/azure.go [hunk 4]
```diff
@@ -317,3 +345,115 @@ func (d *Discovery) refresh() (tg *targetgroup.Group, err error) {
 
 	return tg, nil
 }
+
+func (client *azureClient) getVMs() (vms []virtualMachine, err error) {
```
Comment (brian-brazil) [inline discovery/azure/azure.go]: Can you avoid named return values like this? It's harder to understand the code.

File: discovery/azure/azure.go (hunk 2 above)
Comment (johscheuer) [reply discovery/azure/azure.go]: Fixed in the latest commit.

File: discovery/azure/azure.go (hunk 3 above)
Comment (johscheuer) [reply discovery/azure/azure.go]: result network.Interface,

File: discovery/azure/azure.go (hunk 4 above)
Comment (johscheuer) [reply discovery/azure/azure.go]: I changed this for the other methods as well.

File: discovery/azure/azure.go (hunk 1 above)
Comment (johscheuer) [reply discovery/azure/azure.go]: Actually I don't know why (this is a left over from the initial Azure support) I'm fine with changing this to non-pointer.

File: discovery/azure/azure.go [hunk 5]
```diff
@@ -317,3 +345,120 @@ func (d *Discovery) refresh() (tg *targetgroup.Group, err error) {
...
+			vms = append(vms, mapFromVMScaleSetVM(&vm, scaleSet.Name))
+		}
+	}
+
+	return vms, nil
+}
+
+func mapFromVM(vm *compute.VirtualMachine) virtualMachine {
+	osType := string(vm.Properties.StorageProfile.OsDisk.OsType)
+	return virtualMachine{vm.ID, vm.Name, vm.Type, vm.Location, &osType, nil, vm.Tags, vm.Properties.NetworkProfile}
```
Comment (brian-brazil) [inline discovery/azure/azure.go]: Explicit struct keys are preferred.

File: discovery/azure/azure.go (hunk 1 above)
Comment (brian-brazil) [reply discovery/azure/azure.go]: This appears to be new code, so we may as well clean it up.

Comment (johscheuer) [reply discovery/azure/azure.go]: Fixed in the latest commit

File: discovery/azure/azure.go (hunk 5 above)
Comment (johscheuer) [reply discovery/azure/azure.go]: Fixed in the latest commit

File: discovery/azure/azure.go [hunk 6]
```diff
@@ -45,6 +46,7 @@ const (
 	azureLabelMachineLocation      = azureLabel + "machine_location"
 	azureLabelMachinePrivateIP     = azureLabel + "machine_private_ip"
 	azureLabelMachineTag           = azureLabel + "machine_tag_"
+	azureLabelMachineScaleSet      = azureLabel + "machine_scale_set"
```
Comment (simonpasquier) [inline discovery/azure/azure.go]: This would need to be documented in `docs/configuration/configuration.md`

Comment (johscheuer) [reply discovery/azure/azure.go]: I will create an PR in https://github.com/prometheus/docs, thanks for the hint!

Comment (brian-brazil) [reply discovery/azure/azure.go]: It's all in this repo these days, and should be in this PR

**Comment by brian-brazil** [comment]
This looks okay to me, just needs a docs update.

File: discovery/azure/azure.go (hunk 6 above)
Comment (johscheuer) [reply discovery/azure/azure.go]: Thanks for the hint, I added it.

**Comment by johscheuer** [comment]
Should I squash my commits?

**Comment by brian-brazil** [comment]
Thanks!

No need, GitHub lets me take care of it for you.

**Pull request merged**